### PR TITLE
Upgrading parity-scale-codec to v3

### PR DESCRIPTION
**Pull request by wigy-opensource-developer**
Now that scale-info, primitive-types, frame-metadata, impl-codec and finality-grandpa all have released versions that uses parity-scale-codec v3, this PR should finish integrating it into Substrate.

polkadot companion: paritytech/polkadot#4958
cumulus companion: paritytech/cumulus#1013

### CODE REVIEW

**Comment by wigy-opensource-developer** [comment]
This is how far I got with the upgrade. Turns out the proc-macro decl_runtime_apis generates code that is incompatible with v3, but I did not have the time to fix that up.

File: bin/node-template/pallets/template/Cargo.toml [hunk 1]
```diff
@@ -13,10 +13,12 @@ repository = "https://github.com/substrate-developer-hub/substrate-node-template
 targets = ["x86_64-unknown-linux-gnu"]
 
 [dependencies]
-codec = { package = "parity-scale-codec", version = "2.0.0", default-features = false, features = [
+codec = { package = "parity-scale-codec", version = "3", default-features = false, features = [
+	"derive",
+] }
```
Comment (KiChjang) [inline bin/node-template/pallets/template/Cargo.toml:18]: Can we not apply the Rust coding styles to TOML files? I find them to be highly inappropriate, and someone else almost always comes in and changes them back anyway.

Comment (wigy-opensource-developer) [reply bin/node-template/pallets/template/Cargo.toml:18]: It is a TOML plugin that does this to reduce the line length, I guess. I have no real preference, just hate to cause or do manual work.

**Comment by ordian** [comment]
I don't know how you updated dependencies, but suggest to do the follow:
```
cargo install cargo-edit
cargo upgrade --workspace hashbrown parity-util-mem parity-scale-codec scale-info trie-db memory-db kvdb kvdb-rocksdb kvdb-memorydb primitive-types trie-bench frame-metadata finality-grandpa parking_lot
cargo update -p scale-info:1.0.0
cargo update -p bindgen
```
On another note, one simply does not run `cargo update`! It's too much of a big change to review properly.

**Comment by wigy-opensource-developer** [comment]
> one simply does not run `cargo update`!

Well, in smaller projects I usually did. Do we leave it for github dependabot to upgrade our dependencies? Also, how did you come up with the exact list of dependencies to **upgrade** and the crates to **update**?

I have force pushed the result of those commands, still need to adapt some macros for scale v3.

**Comment by ordian** [comment]
> Do we leave it for github dependabot to upgrade our dependencies?

Doing `cargo update` as a separate PR on polkadot and cumulus with a proper burn-in from time to time makes sense if we don't keep up with dependabot updates.

>  Also, how did you come up with the exact list of dependencies to upgrade and the crates to update?

As a maintainer of parity-common I know that some crates need to upgraded together, e.g. because of [that](https://github.com/paritytech/parity-common/blob/70a745a701f1917dd9c54a009c46d00f6fcb1dee/parity-util-mem/Cargo.toml#L12) (but also in general to avoid duplicates). 
For `cargo update`, I've noticed `scale-info:1.0.0` was not eliminated by `cargo upgrade` alone. Probably because of existing `Cargo.lock` and `>=0.9, <3` version requirement in parity-common.
`cargo update -p bindgen` removes another duplicate dep.

**Comment by ordian** [comment]
A note that this brings https://github.com/paritytech/parity-common/pull/629, which might improve perf (using jemalloc now) for rocksdb.

**Review by ordian — APPROVED** [review]

**Review by bkchr — APPROVED** [review]

File: client/transaction-pool/Cargo.toml [hunk 2]
```diff
@@ -41,7 +41,7 @@ sp-consensus = { version = "0.10.0-dev", path = "../../primitives/consensus/comm
 substrate-test-runtime-transaction-pool = { version = "2.0.0", path = "../../test-utils/runtime/transaction-pool" }
 substrate-test-runtime-client = { version = "2.0.0", path = "../../test-utils/runtime/client" }
 sc-block-builder = { version = "0.10.0-dev", path = "../block-builder" }
-codec = { package = "parity-scale-codec", version = "2.0.0" }
+codec = { package = "parity-scale-codec", version = "3.0.0" }
```
Comment (KiChjang) [inline client/transaction-pool/Cargo.toml:44]: This is weird, `codec` is already listed as a dependency in this file, so there's no need for this to appear again in `dev-dependencies`.

Comment (wigy-opensource-developer) [reply client/transaction-pool/Cargo.toml:44]: Yeah, in some of the cases they are added with different features, but not here. I try to find the history of it.

**Review by KiChjang — COMMENTED** [review]
Just sanity checking that every occurrence of `decode_all` and `decode_all_with_depth_limit` has been dealt with accordingly in the code base, correct? Any `decode_and_advance_with_depth_limit` usages are also fixed as well?

**Comment by wigy-opensource-developer** [comment]
Yes, Substrate build complained loudly and I went through all usages with a manual search. The usages generated by proc-macro were a bit tricky, but were also handled.

**Comment by rcny** [comment]
CI uses the image with `clang-13` now.

**Comment by wigy-opensource-developer** [comment]
> CI uses the image with `clang-13` now.

Perfect, thanks. Now hopefully I just need to fix some tests.

**Pull request merged**